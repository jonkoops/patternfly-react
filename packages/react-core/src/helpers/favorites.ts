import { ReactNode, ReactElement, cloneElement, Children } from 'react';
import { ApplicationLauncherSeparator } from '../deprecated/components/ApplicationLauncher/ApplicationLauncherSeparator';
import { Divider } from '../components/Divider/Divider';

/**
 * This function is a helper for creating an array of renderable favorite items for the Application launcher or Select
 *
 * @param {object} items The items rendered in Select or Application aLauncher
 * @param {boolean} isGrouped Flag indicating if items are grouped
 * @param {any[]} favorites Array of ids of favorited items
 * @param {boolean} isEnterTriggersArrowDown Flag indicating if we should add isEnterTriggersArrowDown to favorited item
 */
export const createRenderableFavorites = (
  items: object,
  isGrouped: boolean,
  favorites: any[],
  isEnterTriggersArrowDown?: boolean
) => {
  if (isGrouped) {
    const favoriteItems: ReactNode[] = [];
    (items as ReactElement[]).forEach((group) => {
      if (favorites.length > 0) {
        return (
          group.props.children &&
          (group.props.children as ReactElement[])
            .filter((item) => favorites.includes(item.props.id))
            .map((item) => {
              if (isEnterTriggersArrowDown) {
                return favoriteItems.push(
                  cloneElement(item, {
                    isFavorite: true,
                    enterTriggersArrowDown: isEnterTriggersArrowDown,
                    id: `favorite-${item.props.id}`
                  })
                );
              } else {
                return favoriteItems.push(cloneElement(item, { isFavorite: true, id: `favorite-${item.props.id}` }));
              }
            })
        );
      }
    });
    return favoriteItems;
  }
  return (items as ReactElement[])
    .filter((item) => favorites.includes(item.props.id))
    .map((item) => cloneElement(item, { isFavorite: true, enterTriggersArrowDown: isEnterTriggersArrowDown }));
};

/**
 * This function is a helper for extending the array of renderable favorite with the select/application launcher items to  render in the Application launcher or Select
 *
 * @param {object} items The items rendered in Select or Application aLauncher
 * @param {boolean} isGrouped Flag indicating if items are grouped
 * @param {any[]} favorites Array of ids of favorited items
 */
export const extendItemsWithFavorite = (items: object, isGrouped: boolean, favorites: any[]) => {
  if (isGrouped) {
    return (items as ReactElement[]).map((group) =>
      cloneElement(group, {
        children: Children.map(group.props.children as ReactElement[], (item) => {
          if (item.type === ApplicationLauncherSeparator || item.type === Divider) {
            return item;
          }
          return cloneElement(item, {
            isFavorite: favorites.some(
              (favoriteId) => favoriteId === item.props.id || `favorite-${favoriteId}` === item.props.id
            )
          });
        })
      })
    );
  }
  return (items as ReactElement[]).map((item) =>
    cloneElement(item, {
      isFavorite: favorites.some((favoriteId) => favoriteId === item.props.id)
    })
  );
};
