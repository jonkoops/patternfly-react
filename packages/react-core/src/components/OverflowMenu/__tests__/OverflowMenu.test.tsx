/**
 * @vitest-environment jsdom
 */
import styles from '@patternfly/react-styles/css/components/OverflowMenu/overflow-menu';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { describe, expect, test, vi } from 'vitest';

import * as getResizeObserver from '../../../helpers/resizeObserver';
import { OverflowMenu } from '../OverflowMenu';

describe('OverflowMenu', () => {
  test('md', () => {
    render(<OverflowMenu breakpoint="md" data-testid="test-id" />);
    expect(screen.getByTestId('test-id')).toHaveClass(styles.overflowMenu);
  });

  test('lg', () => {
    const { asFragment } = render(<OverflowMenu breakpoint="lg" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('xl', () => {
    const { asFragment } = render(<OverflowMenu breakpoint="xl" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('basic', () => {
    const { asFragment } = render(
      <OverflowMenu breakpoint="md">
        <div>BASIC</div>
      </OverflowMenu>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should warn on bad props', () => {
    const myMock = vi.fn() as any;
    global.console = { error: myMock } as any;

    render(
      <OverflowMenu breakpoint={undefined as 'md'}>
        <div>BASIC</div>
      </OverflowMenu>
    );

    expect(myMock).toHaveBeenCalled();
  });

  test('should call resizeObserver on undefined containerRefElement', () => {
    const resizeObserver = vi.spyOn(getResizeObserver, 'getResizeObserver');

    render(<OverflowMenu breakpoint={'md'} />);

    expect(resizeObserver).toHaveBeenCalledWith(undefined, expect.any(Function));
  });

  test('should call resizeObserver on selector ref containerRefElement', () => {
    const resizeObserver = vi.spyOn(getResizeObserver, 'getResizeObserver');

    render(
      <div id="selector-ref-test">
        <OverflowMenu breakpointReference={() => document.getElementById('selector-ref-test')} breakpoint={'md'}>
          <div>Selector ref test</div>
        </OverflowMenu>
      </div>
    );

    const containerRef = screen.getByText('Selector ref test').parentElement?.parentElement;

    expect(resizeObserver).toHaveBeenCalledWith(containerRef, expect.any(Function));
  });

  test('should call resizeObserver on React ref containerRefElement', () => {
    const resizeObserver = vi.spyOn(getResizeObserver, 'getResizeObserver');
    const containerRef = React.createRef<HTMLDivElement>();

    render(
      <div ref={containerRef}>
        <OverflowMenu breakpointReference={containerRef} breakpoint={'md'} />
      </div>
    );

    expect(resizeObserver).toHaveBeenCalledWith(containerRef.current, expect.any(Function));
  });
});
