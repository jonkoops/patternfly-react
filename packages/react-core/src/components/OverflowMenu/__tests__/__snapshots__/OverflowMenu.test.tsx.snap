// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OverflowMenu > basic 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-overflow-menu"
  >
    <div>
      BASIC
    </div>
  </div>
</DocumentFragment>
`;

exports[`OverflowMenu > lg 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-overflow-menu"
  />
</DocumentFragment>
`;

exports[`OverflowMenu > xl 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-overflow-menu"
  />
</DocumentFragment>
`;
